src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/AvoidUninstantiatedInternalClasses.cs

```diff
@@ -109,7 +97,11 @@ public sealed override void Initialize(AnalysisContext context)
                             mef1ExportAttributeSymbol,
                             mef2ExportAttributeSymbol))
                     {
-                        internalTypes.TryAdd(type, null);
+                        if (!hasInternalsVisibleTo
+                            || !context.Options.GetBoolOptionValue(EditorConfigOptionNames.SkipWhenInternalsVisibleIsPresent, Rule, context.Symbol, context.Compilation, defaultValue: true))
```

Comment:
We normally delay the option check to the very last point as this can be comparatively expensive - can cause dictionary lookups within locks and such. I would suggest reverting this change and instead change the CompilationEndAction as follows:

```csharp
                startContext.RegisterCompilationEndAction(context =>
                {
                    var uninstantiatedInternalTypes = internalTypes
                        .Select(it => it.Key.OriginalDefinition)
                        .Except(instantiatedTypes.Select(it => it.Key.OriginalDefinition))
                        .Where(type => !HasInstantiatedNestedType(type, instantiatedTypes.Keys));
                    foreach (var type in uninstantiatedInternalTypes)
                    {
                        if (!hasInternalsVisibleTo
                            || !context.Options.GetBoolOptionValue(EditorConfigOptionNames.SkipWhenInternalsVisibleIsPresent, Rule, type, context.Compilation, defaultValue: true))
                            {
                                context.ReportDiagnostic(type.CreateDiagnostic(Rule, type.FormatMemberName()));
                            }
                    }
                });
```

Reply:
Done.